conn.go

```diff
@@ -87,6 +87,12 @@ type ConnOptions struct {
 
 	// test hook
 	dialer dialer
+
+	// Channel to send close events to the user
+	// the event will be sent when the connection is closed correctly or by an error
+	// the event will contain the error and the connection
+	// if the channel is nil, the events will not be sent
+	ChCloseEvent chan *CloseEvent
```

Comment:
@richardpark-msft and I have talked about such a feature in the past, but it never materialized.

I don't think we want to accept a channel from the caller for two reasons.
- if it's unbuffered it will cause `connReader` to stall until somebody reads from it
- if the caller closes the channel `connReader` will panic

So, then the question becomes, do we want this always on or make it pay-for-play via some kind of "EnableCloseEvent" option. Maybe making it always on isn't that bad?

Reply:
"Always on" seems like a viable option, and the user can block on it similar to what you expect with `context.Done()`. 

I'd also want to add one in for links as well, so we can tell when they detach.

Reply:
Right. So we'd end up with an API like this.
```go
func (c *Conn) Done() <-chan *ConnError {...}
```

Reply:
And similar for senders/receivers.

Reply:
Thank you @jhendrixMSFT @richardpark-msft 
The idea of this PR was to show the concept. 
I agree with https://github.com/Azure/go-amqp/pull/333#discussion_r1761632037 is similar in what there is in .NET amqplite with the event `.AddClosedCallback(_closedCallback)`

Reply:
Understood.

Going through the backlog I found https://github.com/Azure/go-amqp/issues/99 which is when we originally had the idea for such functionality. I've updated it based on this conversation with the pattern discussed above.